### ISSUE TITLE
dflash model seems a lot bigger than eagle3 and not that fast

### ISSUE THREAD

**Issue by brisker** [z-lab/dflash#55]
1. The qwen3_8B dflash model provided [here](https://huggingface.co/z-lab/Qwen3-8B-DFlash-b16/tree/main) is roughly 2.1GB, but for typical qwen3_8B_eagle3 model [here](https://huggingface.co/AngelSlim/Qwen3-8B_eagle3/tree/main), the typical model storage is roughly 800MB, so my question is: why is dflash a lot bigger? If compressed a little bit(for instance, compressed to a 800MB dflash model , will the performance drop significantly, which may be worse than eagle3? Besides, I think the comparison bwtween a 800MB dflash model and a 800MB eagle3 model,**may be more fair.**
2. I  have give it a try, using the two qwen3_8B speculative models **mentioned above**, and evaluate their efficiency using 
https://github.com/sgl-project/SpecForge/blob/main/benchmarks/bench_eagle3.py
and here is some results( using `git+https://github.com/sgl-project/sglang.git@refs/pull/20547/`):

  thoughoutput(token/s)      batchsize=4        batchsize=16      batchsize=64
eagle3-mtbench                        1142                    3549                 5840
dflash-mtbench                          1527                   3844                 4486

  thoughoutput(token/s)        batchsize=4     batchsize=16      batchsize=64
eagle3-gsm8k                             1018                  2311                 3226
dflash-gsm8k                               1315                2574                  2958

the speedup of dflash over eagle3 is well in batchsize=4, not that big in batchsize=16, and in batchsize=64, dflash is even slower..
I have not read the paper very deeply, but my intuition tells me that this may be due to the bigger model size of dflash.
 Can the authors give some discussions?

**Comment by jianc99** [comment]
Thanks for the great questions and for trying out the models!

**1. Why is DFlash larger, and would compression hurt?**

The key insight is that DFlash predicts all draft tokens in parallel, so we can afford a larger (more layers) draft model while still keeping drafting cost lower than a single-layer EAGLE-3, which drafts autoregressively. This is discussed in Section 3 of our paper. As shown in the figure below:

<img width="477" height="267" alt="Image" src="https://github.com/user-attachments/assets/99e23edd-561b-4596-8fc2-91df3c70cf51" />

 A 5-layer DFlash model is faster than 1-layer EAGLE-3 when drafting 4, 8, or 16 tokens — and the gap widens significantly at 16 draft tokens, precisely because EAGLE-3 must generate tokens sequentially.

We do have results comparing a 5-layer EAGLE-3 against a 5-layer DFlash (i.e., same draft model size) on Qwen3-4B with block size 8:

|  | GSM8K AL | GSM8K Spd | Humaneval AL | Humaneval Spd | MT-Bench AL | MT-Bench Spd |
|---|---|---|---|---|---|---|
| EAGLE-3-5L | 4.2 | 2.1x | 4.3 | 2.2x | 3.1 | 1.4x |
| DFlash-5L | 4.2 | 3.3x | 4.0 | 3.2x | 3.0 | 2.2x |

As you can see, DFlash achieves similar acceptance length as 5-layer EAGLE-3 but much higher speedup due to its lower drafting latency. So even at equal size, DFlash wins on speed.

**2. On your throughput benchmarks**

A few things to check:

- **Number of draft tokens**: DFlash's default is 16, and its advantage grows with more draft tokens. Could you confirm how many draft tokens EAGLE-3 is using in your setting? These should be aligned for a fair comparison.
- **Batch size scaling**: In our tests, DFlash's advantage over EAGLE-3 typically grows as batch size increases, when the number of draft tokens is aligned. The pattern you're seeing (DFlash slower at bs=64) may stem from a mismatch in draft token count rather than model size. As larger number of draft tokens will make the verification more costly when batch size is large.
- **Thinking mode**: Our current Qwen3-4B and 8B checkpoints are trained with thinking disabled. If thinking is turned on, performance will degrade. We plan to release retrained checkpoints with thinking enabled soon.

Hope this helps!

**Comment by brisker** [comment]
@jianc99 
1. Could you please provide the scripts to reproduce the results of both eagle3 and dflash ? (including sglang version) . I have aligned the draft token nums  and thinking mode, still, not good results. The models I used are(qwen3_8B) :  
[dflash](https://huggingface.co/z-lab/Qwen3-8B-DFlash-b16/tree/main) and  [eagle3](https://huggingface.co/AngelSlim/Qwen3-8B_eagle3/tree/main)
3. Why not use smaller dflash models, what about the comparision between 1-layer-dflash and 1-layer-eagle3 (eagle3 is originally 1-layer,right?) ? 
(Since the storage could also be **expensive**, if applied into small-devices like phones, models could be small, and 5-layer and 1-layer may not be that small thing)
5. When will the training code be published?
6. Since more draft token num in eagle3 may lead to worse results, at least in many eagle3 experiments in my experience, **4 draft tokens may be the best choice,** and 16 draft tokens may be much worse, But to compare with dflash, you just simply set the draft token num as 16 for eagle3? **Will this point cause some unfair comparision?**

**Comment by jianc99** [comment]
1. You can install the SGLang dependencies in this repo by following the installation instructions, then serve two Qwen3 SGLang servers:

```bash
python -m sglang.launch_server \
    --model-path Qwen/Qwen3-8B \
    --speculative-algorithm EAGLE3 \
    --speculative-draft-model-path AngelSlim/Qwen3-8B_eagle3 \
    --speculative-num-draft-tokens 16 \
    --speculative-num-steps 15 \
    --speculative-eagle-topk 1 \
    --tp-size 1 \
    --attention-backend flashinfer \
    --mem-fraction-static 0.75 \
    --trust-remote-code
```

```bash
python -m sglang.launch_server \
    --model-path Qwen/Qwen3-8B \
    --speculative-algorithm DFLASH \
    --speculative-draft-model-path z-lab/Qwen3-8B-DFlash-b16 \
    --speculative-num-draft-tokens 16 \
    --tp-size 1 \
    --attention-backend flashinfer \
    --mem-fraction-static 0.75 \
    --trust-remote-code
```

Then run the following benchmark script against each server:

```bash
python -m dflash.benchmark --backend sglang \
    --base-url http://127.0.0.1:30000 --model Qwen/Qwen3-8B \
    --dataset gsm8k --num-prompts 1024 --concurrency 32
```

Below are the results from my end on 1×B200:

EAGLE-3:

```bash
==================================================
Backend:          sglang
Dataset:          gsm8k
Num prompts:      1024
Concurrency:      32
Latency:          77.7s
Output tokens:    310577
Throughput:       3,995.90 tok/s
Accept length:    2.575
Spec verify ct:   121387
==================================================
```

DFlash:

```bash
==================================================
Backend:          sglang
Dataset:          gsm8k
Num prompts:      1024
Concurrency:      32
Latency:          34.3s
Output tokens:    309504
Throughput:       9,018.70 tok/s
Accept length:    6.469
Spec verify ct:   49445
==================================================
```

As you can see, DFlash achieves significantly higher throughput and acceptance length. Feel free to try different concurrency levels, block sizes or EAGLE-3 tree settings to compare them further.

2. The reason we use a 5-layer DFlash draft model is that, as discussed in Section 5.4.2 and Table 5 of the paper, a 5-layer drafter achieves the best trade-off between drafting cost and acceptance length, delivering the highest end-to-end speedup compared to both shallower and deeper drafters.

<img width="573" height="239" alt="Image" src="https://github.com/user-attachments/assets/3a82a44b-e3fc-414c-b716-18227581fa3e" />

3. We will release the training code soon. In the meantime, there are already some open-source training resources available for DFlash, for example, SpecForge, Speculators, and Model-Optimizer all have relevant PRs or have already supported DFlash training.

4. In the paper, we compare EAGLE-3 under two settings: (1) 7 draft steps, top-k=10, and 60 draft tokens, which aligns with the EAGLE-3 paper's original setting; and (2) 7 draft steps, top-k=10, with 16 or 10 draft tokens, which aligns the draft token count with DFlash. This ensures a fair comparison. On the Qwen3.5 series HuggingFace model pages, we also compare block size 8 and 16 results against Qwen3.5's native MTP (which is similar to EAGLE). You can adjust the number of draft tokens further if needed to compare DFlash and EAGLE-3.
